homeassistant/components/folder_watcher.py

```diff
@@ -0,0 +1,85 @@
+"""
+Component for monitoring activity on a folder.
+"""
+import os
+import logging
+import voluptuous as vol
+from watchdog.observers import Observer
+from watchdog.events import FileSystemEventHandler
+
+from homeassistant.helpers.entity import Entity
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['watchdog']
+_LOGGER = logging.getLogger(__name__)
+
+CONF_FOLDERS = 'folders'
+DOMAIN = "folder_watcher"
+DEFAULT_IS_RECURSIVE = True
+EVENT_TYPE = "event_type"
+IS_DIRECTORY = "is_directory"
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_FOLDERS):
+            vol.All(cv.ensure_list, [cv.isdir]),
+        })
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, config):
+    """Set up the folder watcher."""
+    conf = config[DOMAIN]
+    paths = conf[CONF_FOLDERS]
+    for path in paths:
+        if not hass.config.is_allowed_path(path):
+            _LOGGER.error("folder %s is not valid or allowed", path)
+            return False
+        else:
+            Watcher(path, hass)
+    return True
+
+
+class Watcher(Entity):
+    """Class for starting Watchdog."""
+    def __init__(self, path, hass):
+        self._observer = Observer()
+        self._observer.schedule(
+            MyHandler(hass),
+            path,
+            recursive=DEFAULT_IS_RECURSIVE)
+        self._observer.start()
```

Comment:
How to listen for HOMEASSISTANT_START - I have to apply a decorator to `self._observer.start()`?

Reply:
I didn't really understand @balloob's comment from before, since from what I know this *is* an entity. Here's how I would have done it:

```python
class FolderWatcher(Entity, FileSystemEventHandler):
    def __init__(self, path):
        self._observer = Observer()
        self._path = path

    async def async_added_to_hass(self):
        self.hass.async_add_job(self.start_watching)

    async def async_will_remove_from_hass(self):
        self.hass.async_add_job(self.stop_watching)

    def start_watching(self):
        self._observer.schedule(self, self._path, recursive=True)
        self._observer.start()

    def stop_watching(self):
        self._observer.stop()

    def process(self, event):
        on_any_event(self.hass, event)

    def on_modified(self, event):
        self.process(event)

    def on_moved(self, event):
        self.process(event)

    def on_created(self, event):
        self.process(event)

    def on_deleted(self, event):
        self.process(event)
```

Reply:
(My GitHub client is not working correctly, repost)

I didn't really understand @balloob's comment from before, since from what I know this *is* an entity. Here's how I would have done it:

```python
class FolderWatcher(Entity, FileSystemEventHandler):
    def __init__(self, path):
        self._observer = Observer()
        self._path = path

    async def async_added_to_hass(self):
        self.hass.async_add_job(self.start_watching)

    async def async_will_remove_from_hass(self):
        self.hass.async_add_job(self.stop_watching)

    def start_watching(self):
        self._observer.schedule(self, self._path, recursive=True)
        self._observer.start()

    def stop_watching(self):
        self._observer.stop()

    def process(self, event):
        on_any_event(self.hass, event)

    def on_modified(self, event):
        self.process(event)

    def on_moved(self, event):
        self.process(event)

    def on_created(self, event):
        self.process(event)

    def on_deleted(self, event):
        self.process(event)
```

Reply:
@OttoWinter tried that but async_added_to_hass() is never called

Reply:
I also receive another error using `FolderWatcher(Entity, FileSystemEventHandler):` that I don't understand:
```
2018-03-08 06:36:30 ERROR (MainThread) [homeassistant.setup] Error during setup of component folder_watcher
Traceback (most recent call last):
  File "/Users/robincole/Documents/Github/home-assistant/homeassistant/setup.py", line 145, in _async_setup_component
    component.setup, hass, processed_config)
  File "/anaconda3/lib/python3.6/asyncio/futures.py", line 332, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/anaconda3/lib/python3.6/asyncio/tasks.py", line 250, in _wakeup
    future.result()
  File "/anaconda3/lib/python3.6/asyncio/futures.py", line 245, in result
    raise self._exception
  File "/anaconda3/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/robincole/.homeassistant/custom_components/folder_watcher.py", line 38, in setup
    FolderWatcher(path)
  File "/Users/robincole/.homeassistant/custom_components/folder_watcher.py", line 57, in __init__
    self._observer.schedule(self, self._path, recursive=True)
  File "/Users/robincole/Documents/Github/home-assistant/lib/python3.6/site-packages/watchdog/observers/fsevents.py", line 172, in schedule
    return BaseObserver.schedule(self, event_handler, path, recursive)
  File "/Users/robincole/Documents/Github/home-assistant/lib/python3.6/site-packages/watchdog/observers/api.py", line 284, in schedule
    self._add_handler_for_watch(event_handler, watch)
  File "/Users/robincole/Documents/Github/home-assistant/lib/python3.6/site-packages/watchdog/observers/api.py", line 243, in _add_handler_for_watch
    self._handlers[watch].add(event_handler)
TypeError: unhashable type: 'FolderWatcher'
```

Reply:
Are you using the `setup_`**`platform()`** code? I'm pretty sure `async_added_to_hass` would be called for entities of a platform component and quite a few integrations depend on it.

```python
def setup_platform(hass, config, add_devices, discovery_info=None):
    # ...
    add_devices([FolderWatcher(...)])
```

That *should* call the `async_added_to_hass` callback.

Reply:
Oh the unhashable type is an issue. If I understand correctly `watchdog` tries to put the entity in a `set`. And to be added to a set an object must be hashable (implementing the hash function). Entities however are apparently unhashable, so using `class FolderWatcher(Entity, FileSystemEventHandler):` won't work.

Reply:
@OttoWinter I'm using setup(), also I am confused now why you recommend setup_platform() as this is a component, not a sensor, perhaps I'm missing something?

Reply:
Ohhhh! Sorry I thought this would be a replacement for the folder sensor, just with the added benefit of firing events. However, you correctly said "Replaces #12912 as becomes a standalone component." at the top of your PR.

Reply:
If you want to learn how to use the start event, check the code… `git grep EVENT_HOMEASSISTANT_START`